### PR TITLE
Adding tests for the ML pipelines API

### PR DESCRIPTION
**Pull request by DharmitD**

### CODE REVIEW

File: tests/basictests/ml-pipelines.sh [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    os::cmd::try_until_text "oc get pods -l application-crd-id=kubeflow-pipelines --field-selector='status.phase=Running' -o jsonpath='{$.items[*].metadata.name}' | wc -w" "0" $odhdefaulttimeout $odhdefaultinterval
+}
+
+check_resources
+check_ui_overlay
+check_pipeline_api
+teardown_pipelines
+
+
+os::test::junit::declare_suite_end
```
Comment (rimolive) [inline tests/basictests/ml-pipelines.sh:66]: New line

File: tests/setup/kfctl_openshift.yaml [hunk 2]
```diff
@@ -169,9 +169,19 @@ spec:
...
+    name: kfp-tekton
   repos:
   - name: kf-manifests
     uri: https://github.com/opendatahub-io/manifests/tarball/master
   - name: manifests
     uri: https://github.com/opendatahub-io/odh-manifests/tarball/master
-  version: master
+  - name: pipeline-manifests
+    uri: https://github.com/opendatahub-io/odh-manifests/tarball/pipelines-prototype
+  version: master
```
Comment (rimolive) [inline tests/setup/kfctl_openshift.yaml:187]: New line

File: tests/basictests/ml-pipelines.sh [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+    SVC=$(oc get route ml-pipeline --template={{.spec.host}})
+    echo "Creating a pipeline"
+    PIPELINE_ID=$(curl -F "uploadfile=@${RESOURCEDIR}/ml-pipelines/kfp-tekton.yaml" ${SVC}/apis/v1beta1/pipelines/upload | jq -r .id)
+    os::cmd::try_until_text "curl ${SVC}/apis/v1beta1/pipelines/${PIPELINE_ID} | jq '.name'" "kfp-tekton.yaml" $odhdefaulttimeout $odhdefaultinterval
+
+    echo "Listing pipelines"
+    os::cmd::try_until_text "curl ${SVC}/apis/v1beta1/pipelines | jq '.total_size'" "1" $odhdefaulttimeout $odhdefaultinterval #change to one
+
+   echo "Create a run"
```
Comment (rimolive) [inline tests/basictests/ml-pipelines.sh:34]: wrong identation

File: tests/setup/kfctl_openshift.yaml [hunk 4]
```diff
@@ -169,9 +169,19 @@ spec:
         name: manifests
         path: model-mesh
     name: model-mesh
+  - kustomizeConfig:
+      overlays:
+        - metadata-store-mysql
+        - ml-pipeline-ui
```
Comment (rimolive) [inline tests/setup/kfctl_openshift.yaml:175]: I don't think we should discuss if we want to test `ml-pipeline-ui` overlay here. As it is just basic tests, we should test basic flows, and the UI seems to me like a special case.

File: tests/resources/ml-pipelines/kfp-tekton.yaml [hunk 5]
```diff
@@ -0,0 +1,639 @@
...
+            try:
+              input1=int(input1)
+              input2=int(input2)
+            except:
+              input1=str(input1)
+            outcome="true" if (input1 $(inputs.params.operator) input2) else "false"
+            f = open("/tekton/results/outcome", "w")
+            f.write(outcome)
+            f.close()' '$(inputs.params.operand1)' '$(inputs.params.operand2)'
+          image: quay.io/rmartine/python-36:latest
```
Comment (rimolive) [inline tests/resources/ml-pipelines/kfp-tekton.yaml:597]: Better move this image under odh repo in quay.io

**Review by rimolive — CHANGES_REQUESTED** [review]

**Comment by DharmitD** [comment]
Also segregated each of the curl commands into a separate function as discussed. Tested all the changes above on a clean cluster and the changes work as expected.

File: tests/basictests/ml-pipelines.sh [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+function check_ui_overlay() {
+    header "Checking UI overlay Kfdef deploys the UI"
+    os::cmd::try_until_text "oc get pods -l app=ml-pipeline-ui --field-selector='status.phase=Running' -o jsonpath='{$.items[*].metadata.name}' | wc -w" "1" $odhdefaulttimeout $odhdefaultinterval
+}
+
+function create_pipeline() {
+    header "Creating a pipeline"
+    SVC=$(oc get route ml-pipeline --template={{.spec.host}})
+    PIPELINE_ID=$(curl -F "uploadfile=@${RESOURCEDIR}/ml-pipelines/kfp-tekton.yaml" ${SVC}/apis/v1beta1/pipelines/upload | jq -r .id)
+    os::cmd::try_until_text "curl ${SVC}/apis/v1beta1/pipelines/${PIPELINE_ID} | jq '.name'" "kfp-tekton.yaml" $odhdefaulttimeout $odhdefaultinterval
```
Comment (VaishnaviHire) [inline tests/basictests/ml-pipelines.sh:27]: Won't we need to prepend `https://` before ${SVC} ?

File: tests/basictests/ml-pipelines.sh [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    header "Listing runs"
+    os::cmd::try_until_text "curl ${SVC}/apis/v1beta1/runs | jq '.total_size'" "1" $odhdefaulttimeout $odhdefaultinterval
+}
+
+function check_run_status() {
+    header "Checking run status"
+    os::cmd::try_until_text "curl ${SVC}/apis/v1beta1/runs/${RUN_ID} | jq '.run.status'" "Completed" $odhdefaulttimeout $odhdefaultinterval
+}
+
+function delete_runs() {
```
Comment (VaishnaviHire) [inline tests/basictests/ml-pipelines.sh:50]: Can we also test deletion of resources, similar to what we have in other tests ?
https://github.com/opendatahub-io/odh-manifests/blob/master/tests/basictests/odh-notebook-controller.sh#L84

File: tests/basictests/ml-pipelines.sh [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    header "Deleting runs"
+    os::cmd::try_until_text "curl -X DELETE ${SVC}/apis/v1beta1/runs/${RUN_ID} | jq" "" $odhdefaulttimeout $odhdefaultinterval
+    os::cmd::try_until_text "curl ${SVC}/apis/v1beta1/runs/${RUN_ID} | jq '.code'" "5" $odhdefaulttimeout $odhdefaultinterval
+}
+
+function delete_pipeline() {
```
Comment (VaishnaviHire) [inline tests/basictests/ml-pipelines.sh:56]: Same as above

File: tests/basictests/ml-pipelines.sh (hunk 7 above)
Comment (DharmitD) [reply tests/basictests/ml-pipelines.sh:50]: As per discussion with @VaishnaviHire, added a function to check that the pipeline resources are cleaned up.

File: tests/basictests/ml-pipelines.sh (hunk 8 above)
Comment (DharmitD) [reply tests/basictests/ml-pipelines.sh:56]: As per discussion with @VaishnaviHire, added a function to check that the pipeline resources are cleaned up.

**Review by anishasthana — APPROVED** [review]

**Pull request merged**